### PR TITLE
Command: Generate file path for test command

### PR DESCRIPTION
**Pull request by abeatrix**
Fix https://github.com/sourcegraph/cody/issues/3311

- Updated the generic prompt rules to anticipate a file path for the test suite.
- Modified the edit provider to achieve the following:
- Extract the full file path instead of just the file name.
- Utilize workspace URI, if available, to resolve the full path.
- Default to the current directory if no workspace is open.

This adjustment enables the LLM to propose paths relative to the workspace root (as included in the shared context) for the generated test code. This enhancement ensures that the test files can be created in the correct directory, rather than always being saved in the same directory as the current file. This is particularly beneficial as it aligns with the intended location for test files.

### Before

The newly generated test file will be saved at the same level as the current file even when there is an existing test file with the same name located in a different directory:

image

https://github.com/sourcegraph/cody/assets/68532117/10cf9341-ca67-41da-9058-9e5c950c9438

### After

The same test that was generated by the LLM is now magically appended to the current file that lives outside the current directory

image

https://github.com/sourcegraph/cody/assets/68532117/123d88c2-f7f9-4b90-9bf6-ec15b1257a97

### CODE REVIEW

File: vscode/src/commands/context/unit-test-file.ts [hunk 1]
```diff
@@ -23,20 +23,24 @@ export async function getContextFilesForUnitTestCommand(file: URI): Promise<Cont
...
         const max = 10
 
-        // Search for test files in the current directory first
-        const curerntDirPattern = getSearchPatternForTestFiles(file, searchInCurrentDirectoryOnly)
+        // Search for a test file that has the same file name first
+        const sameNameTestPattern = getSearchPatternForTestFiles(file, !searchCurrentDirectoryOnly, true)
+        const testWithSameName = await getWorkspaceFilesContext(sameNameTestPattern, excludePattern, 1)
+
+        // Search for test files in the current directory
+        const curerntDirPattern = getSearchPatternForTestFiles(file, searchCurrentDirectoryOnly)
```
Comment (umpox) [inline vscode/src/commands/context/unit-test-file.ts:35]:
```suggestion
        const currentDirPattern = getSearchPatternForTestFiles(file, searchCurrentDirectoryOnly)
```

nit: typo

File: vscode/src/commands/context/unit-test-file.ts [hunk 2]
```diff
@@ -23,20 +23,24 @@ export async function getContextFilesForUnitTestCommand(file: URI): Promise<Cont
...
         // The max number of files to search for in each workspace search
         const max = 10
 
-        // Search for test files in the current directory first
-        const curerntDirPattern = getSearchPatternForTestFiles(file, searchInCurrentDirectoryOnly)
+        // Search for a test file that has the same file name first
+        const sameNameTestPattern = getSearchPatternForTestFiles(file, !searchCurrentDirectoryOnly, true)
+        const testWithSameName = await getWorkspaceFilesContext(sameNameTestPattern, excludePattern, 1)
+
+        // Search for test files in the current directory
```
Comment (umpox) [inline vscode/src/commands/context/unit-test-file.ts:34]: If we found a test file with the same name, is there any point looking for other test files?

File: vscode/src/commands/context/unit-test-file.ts [hunk 3]
```diff
@@ -23,20 +23,24 @@ export async function getContextFilesForUnitTestCommand(file: URI): Promise<Cont
...
+        const curerntDirPattern = getSearchPatternForTestFiles(file, searchCurrentDirectoryOnly)
         const currentDirContext = await getWorkspaceFilesContext(curerntDirPattern, excludePattern, max)
 
-        contextFiles.push(...currentDirContext)
+        contextFiles.push(...testWithSameName, ...currentDirContext)
 
         // If no test files found in the current directory, search the entire workspace
-        if (!contextFiles.length) {
-            const wsTestPattern = getSearchPatternForTestFiles(file, !searchInCurrentDirectoryOnly)
+        if (contextFiles.length < 3) {
```
Comment (umpox) [inline vscode/src/commands/context/unit-test-file.ts:41]: We probably need to update this comment to explain why `3`, does this have a noticeable benefit over `!contextFiles.length`?

File: vscode/src/edit/prompt/context.ts [hunk 4]
```diff
@@ -47,6 +47,7 @@ const getContextFromIntent = async ({
          * The preceding code is already included as part of the response to better guide the output.
          */
         case 'test':
+            return []
```
Comment (umpox) [inline vscode/src/edit/prompt/context.ts:50]: The doc string should still be attached to case 'add` here, but ignore me if you're still playing around with this.

I'm curious if this shouldn't use the same context as `add` though, as if we're appending to an existing test file then it seems useful that we include the surrounding code? Maybe we should just include the preceding code for test, as we'll always insert at the bottom?

**Review by umpox — COMMENTED** [review]
Nice! Left some small comments

File: vscode/src/commands/context/unit-test-file.ts (hunk 3 above)
Comment (abeatrix) [reply vscode/src/commands/context/unit-test-file.ts:41]: ah this was a left over, will remove it!

File: vscode/src/edit/prompt/context.ts (hunk 4 above)
Comment (abeatrix) [reply vscode/src/edit/prompt/context.ts:50]: I removed it because we already add it when we generate the context for test, so the `add` was adding duplicated context in this case. Also, we are appending to an existing test file but the doc file it reads here is the origin file which is a non-test file (if this makes sense)

File: vscode/src/commands/context/unit-test-file.ts (hunk 2 above)
Comment (abeatrix) [reply vscode/src/commands/context/unit-test-file.ts:34]: I'd say yes as it'd be helpful to use other test files as references in terms of patterns and library usage. It also helps in cases where the test file with the same name is empty or short that's unhelpful 😄

File: vscode/src/edit/prompt/context.ts (hunk 4 above)
Comment (umpox) [reply vscode/src/edit/prompt/context.ts:51]: 👍  Can we add a docstring explaining this and move the docstring that's existing to above `case 'add':`

File: vscode/src/commands/context/unit-test-file.ts (hunk 3 above)
Comment (umpox) [reply vscode/src/commands/context/unit-test-file.ts:41]: does this still need removing?

**Review by umpox — APPROVED** [review]
Approved, left a few comments that need changing above
